msg/output_control.msg

```diff
@@ -0,0 +1,95 @@
+uint64 timestamp		 # time since system start (microseconds)
+uint64 timestamp_sample	 # the timestamp the data this control response is based on was sampled
+
+uint8 NUM_OUTPUT_CONTROL_GROUPS = 4 # Max number of instances of this topic
+uint8 MAX_ACTUATORS = 16  # Max actuators available in a single topic publication
+
+uint8 n_outputs
+uint16[16] function       # Actuator function being controlled
```

Comment:
I am curious why we need this ? 
I would think topic ID + value index is sufficient to know which function is being set.
Ex: output_control_ca.value[0] is function FUNCTION_CA0.

Reply:
Other functions are things like MAVLink-controlled servos, internally-controlled payload actuators or landing gear, nose wheel, ...
This now gives you the ability to define a function for your specific needs (for non-flight-control actuators) and control it from anywhere inside PX4.  For another example, the one of the FW controllers (or Navigator) could command the landing gear to deploy at a certain phase of a landing sequence, without having to "hack" the landing gear into the mixer file or actuator effectiveness matrix.

Reply:
> This now gives you the ability to define a function for your specific needs

Sure, and that is great. Simply I think only the mixer should know about the function enum. 
Ex: let's say the FW controller publishes its ailerons setpoints on its own topic. Then, in the mixer module, if an output is configured for "left aileron" the mixer knows that it has to subscribe to the topic output_control_fw and read value number 2. The fixed wing controller should not repeat this config in it's publication otherwise that would be redundant and may create conflict. With my suggestion you have an unambiguous mapping between function enum and {topic+index}.

Reply:
BTW, the different output topics do not need to share the same format as you did. 

For example the gear function may have its dedicated topic with explicit names such as gear_leg, door, brake, lights. No need to "hack" it in an array of 16 floats. Only the mixer should know how to map function enum to a topic field

Reply:
> Ex: let's say the FW controller publishes its ailerons setpoints on its own topic. Then, in the mixer module, if an output is configured for "left aileron" the mixer knows that it has to subscribe to the topic output_control_fw and read value number 2.

This is **exactly** what I want to avoid - mapping functions implicitly from an index in a topic is NOT scalable.  Let's say we start out with 1 landing_gear index, then someone needs a 2nd one down the road.  Do we just keep adding indices to a topic to add new functions?  And what happens when two implicitly-defined functions inside of one topic need to be controlled from two different modules?  

> For example the gear function may have its dedicated topic with explicit names such as gear_leg, door, brake, lights. No need to "hack" it in an array of 16 floats. Only the mixer should know how to map function enum to a topic field

But we're not using the mixer anymore, and I don't want this to balloon to a crazy number of uORB topics doing essentially the same thing if we can put a more general method in place.

I want any module within PX4 to be able to control some sort of "simple" output that is related to it, whether it be landing gear, a payload release servo, a docking station latch, a nose wheel, a PWM-controlled LED, etc., and I want to allow developers to easily add their own custom functions with essentially zero overhead.

Reply:
Think of this PR as the equivalent of using MAVLink `COMMAND_LONG` vs. a bunch of individual messages that must be maintained independently.

Reply:
> This is exactly what I want to avoid - mapping functions implicitly from an index in a topic is NOT scalable.

I think we do not understand each other. PX4 uses different topics to group related data (rate setpoint, gyro data, etc). This is very scalable, and very developper friendly because topic fields have meaningful names. 
The outlier was actuator_controls, which packed data in float arrays with implicit meaning. This was not scalable, the array elements had implicit meaning, several copies of the topic had to be used, etc. This is what we want to avoid and one of the reasons why a mixer refactor is needed.
By using a generic topic for output_control, you again obscure the content of the topic, and require several copies of the topic to allow publication from several modules, this replicates some flaws of the current mixer.
What I suggest is to have a dedicated topic for each function (ex: landing gear), or each group of functions (ex: control allocation outputs, rc channels). This is the way the rest of PX4 works. Then, the content of each topic is explicit. Which module is responsible for which function becomes cristal clear. And more importantly this is more scalable.

>  Let's say we start out with 1 landing_gear index, then someone needs a 2nd one down the road. Do we just keep adding indices to a topic to add new functions? 

We either add a field in an existing topic, or create a new topic depending on what makes most sense.

>  And what happens when two implicitly-defined functions inside of one topic need to be controlled from two different modules?

I would argue this is the sign of a poorly designed system, a missing level of abstraction, probably fixed with a module responsible for the publication of said topic.
(Btw I do not suggest 'implicit' anything)

> But we're not using the mixer anymore, and I don't want this to balloon to a crazy number of uORB topics doing essentially the same thing if we can put a more general method in place.

Exactly, your design requires several copies of the same topic, at least one per module willing to write to outputs. Here I do not think 'general' is helpful. We need specific, dead-simple output functions.

> I want any module within PX4 to be able to control some sort of "simple" output that is related to it, whether it be landing gear, a payload release servo, a docking station latch, a nose wheel, a PWM-controlled LED, etc.

On that we agree :-)

> I want to allow developers to easily add their own custom functions with essentially zero overhead.

With my suggestion, adding a new output function takes a new subscription in mixer_modules (if the new function cannot be added to an existing topic), one new case in a switch statement in mixer_modules, and one more element in the function enum.

Reply:
Ok ok, I understand now!  I still prefer the key:value approach, but I can see some of the benefits of this approach.  I'll have to think about this some more.  Don't forget you also need to add all of these to the logger, too, as separate topics to log (which adds some overhead).

@dagar, any opinions on the matter?